doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -1,209 +1,213 @@
-        that succeeds if at least one of these two tactics succeeds.
+   where :g:`EQ` is the Leibniz equality `eq` or a registered setoid equality.  (Typically
+   the infix notation :n:`@term__1 = @term__2` is defined as :n:`eq @term__1 @term__2`.
+   Either form may used.)
 
-      Orientation :g:`->` or :g:`<-` can be inserted before the :token:`term` to rewrite.
+   .. todo PR: I can't remember where the following sentence came from.  It seems
+      incorrect.  Maybe worth having some comparison to setoid_rewrite, though
 
-   .. tacv:: rewrite @term at @occurrences
+      Use :tacn:`setoid_rewrite`, described in :ref:`generalizedrewriting`, when
+      `eq` is a setoid equality (i.e. any equivalence relation).
 
-      Rewrite only the given :token:`occurrences` of :token:`term`. Occurrences are
-      specified from left to right as for pattern (:tacn:`pattern`). The rewrite is
-      always performed using setoid rewriting, even for Leibniz’s equality, so one
-      has to ``Import Setoid`` to use this variant.
+   :n:`rewrite @one_term` finds subterms matching :n:`@term__1` in the goal,
+   and replaces them with :n:`@term__2` (or the reverse if `<-` is given).
+   Some of the variables :g:`x`\ :sub:`i` are solved by unification,
+   and some of the types :n:`A__1, ..., A__n` may become new
+   subgoals.  It won't find occurrences inside `forall`; use :tacn:`setoid_rewrite`
+   if you want to find such occurrences.
 
-   .. tacv:: rewrite @term by @tactic
+   :n:`{+, @oriented_rewriter }`
+     The :n:`@oriented_rewriter`\s are applied sequentially
+     to the first goal generated by the previous :n:`@oriented_rewriter`.  If any of them fail,
+     the tactic fails.
 
-      Use tactic to completely solve the side-conditions arising from the
-      :tacn:`rewrite`.
+   :n:`{? {| -> | <- } }`
+     For `->` (the default), :n:`term__1` is rewritten
+     into :n:`term__2`.  For `<-`, :n:`term__2` is rewritten into :n:`term__1`.
 
-   .. tacv:: rewrite {+, @orientation @term} {? in @ident }
+   :n:`{? @natural } {? {| ? | ! } }`
+     :n:`natural` is the number of rewrites to perform.  If `?` is given, :n:`@natural`
+     is the maximum number of rewrites to perform; otherwise :n:`natural` is the exact number
+     of rewrites to perform.
 
-      Is equivalent to the `n` successive tactics :n:`{+; rewrite @term}`, each one
-      working on the first subgoal generated by the previous one. An :production:`orientation`
-      ``->`` or ``<-`` can be inserted before each :token:`term` to rewrite.  One
-      unique clause can be added at the end after the keyword in; it will then
-      affect all rewrite operations.
+     '?' performs the rewrite as many times as possible (possibly zero times).
+     This form never fails.  '!' performs the rewrite as many times as possible
+     and at least once.  The tactic fails if the requested number of rewrites can't
+     be performed.
```

Comment:
Maybe clarify that this is when these symbols are used alone (without `natural`).